runtime/src/utils/buffer/append.rs

```diff
@@ -134,6 +161,30 @@ impl<B: Blob> Append<B> {
         Ok(())
     }
 
+    /// Convert this [Append] wrapper to an [Immutable] wrapper.
+    ///
+    /// The caller *must* sync the [Append] before calling this conversion,
+    /// otherwise any unflushed data in the write buffer will be lost.
+    pub async fn into_immutable(self) -> Immutable<B> {
+        // After sync the append buffer contains only the trailing bytes.
+        // Take ownership of the append buffer if possible to avoid cloning.
+        let (blob_size, trailing) = match Arc::try_unwrap(self.buffer) {
+            Ok(buffer) => {
+                let (mut buffer, blob_size) = buffer.into_inner();
+                let trailing_size = blob_size.saturating_sub(buffer.offset) as usize;
+                buffer.data.truncate(trailing_size);
+                (blob_size, buffer.data)
+            }
+            Err(buffer) => {
+                let (buffer, blob_size) = &*buffer.read().await;
+                let trailing_size = blob_size.saturating_sub(buffer.offset) as usize;
+                (*blob_size, buffer.data[..trailing_size].to_vec())
+            }
+        };
```

Comment:
This might be unnecessarily complex, let me know if you think it's not worth it.